staging/src/k8s.io/apiserver/pkg/endpoints/filters/constrained_impersonation.go

```diff
@@ -0,0 +1,375 @@
+		}
+	}
+
+	if (hasGroups || hasUserExtra || hasUID) && !hasUser {
+		return nil, nil, fmt.Errorf("requested %v without impersonating a user", attrsList)
+	}
+
+	if requestedUser != user.Anonymous {
+		// When impersonating a non-anonymous user, include the 'system:authenticated' group
+		// in the impersonated user info:
+		// - if no groups were specified
+		// - if a group has been specified other than 'system:authenticated'
+		//
+		// If 'system:unauthenticated' group has been specified we should not include
+		// the 'system:authenticated' group.
+		addAuthenticated := true
+		for _, group := range groups {
+			if group == user.AllAuthenticated || group == user.AllUnauthenticated {
+				addAuthenticated = false
+				break
+			}
+		}
+
+		if addAuthenticated {
+			groups = append(groups, user.AllAuthenticated)
+		}
+	} else {
+		addUnauthenticated := true
+		for _, group := range groups {
+			if group == user.AllUnauthenticated {
+				addUnauthenticated = false
+				break
+			}
+		}
+
+		if addUnauthenticated {
+			groups = append(groups, user.AllUnauthenticated)
+		}
+	}
+
+	newUser := &user.DefaultInfo{
+		Name:   requestedUser,
+		Groups: groups,
+		Extra:  userExtra,
+		UID:    uid,
+	}
+
+	return attrsList, newUser, nil
+}
+
+func newAttributeRecord(requestor user.Info) *authorizer.AttributesRecord {
+	return &authorizer.AttributesRecord{
+		APIGroup:        authenticationv1.SchemeGroupVersion.Group,
+		APIVersion:      authenticationv1.SchemeGroupVersion.Version,
+		User:            requestor,
+		ResourceRequest: true,
+	}
+}
+
+func authorizeConstrainedImpersonation(ctx context.Context, targetAttributesList []*authorizer.AttributesRecord, a authorizer.Authorizer) (authorizer.Decision, string, authorizer.Attributes, error) {
+	// Get request attributes from the context.
+	requestorAttrs, err := GetAuthorizerAttributes(ctx)
+	if err != nil {
+		return authorizer.DecisionNoOpinion, "", &authorizer.AttributesRecord{}, err
+	}
+
+	for _, actingAsAttributes := range targetAttributesList {
+		decision, reason, err := a.Authorize(ctx, actingAsAttributes)
+
+		if decision == authorizer.DecisionAllow {
+			continue
+		}
+
+		// Fallback to authorize impersonate:node
+		if actingAsAttributes.Verb == impersonateScheduledNodeVerb {
+			attrCopy := copyAuthorizerAttr(actingAsAttributes)
+			attrCopy.Verb = impersonateNodeVerb
+			decision, reason, err = a.Authorize(ctx, attrCopy)
+		}
+
+		if err != nil || decision != authorizer.DecisionAllow {
+			return decision, reason, actingAsAttributes, err
+		}
+	}
+
+	// Prepend the impersonate-on prefix to the actual verb.
+	impersonateOnAttrs := copyAuthorizerAttr(requestorAttrs)
+	impersonateOnAttrs.Verb = fmt.Sprintf("%s%s", impersonateOnVerbPrefix, impersonateOnAttrs.Verb)
+	decision, reason, err := a.Authorize(ctx, impersonateOnAttrs)
+	return decision, reason, impersonateOnAttrs, err
+}
+
+func copyAuthorizerAttr(attr authorizer.Attributes) *authorizer.AttributesRecord {
```

Comment:
I haven't checked yet, but this needs a fuzzer test to be sure that new fields will cause the copy to fail until the function is updated